Actor.cpp

```diff
@@ -1,142 +1,83 @@
+
+	char fileName[MAX_PATH];
+	char writerFoundEvent[MAX_PATH];
+	char readerFinishedEvent[MAX_PATH];
+	sprintf_s(fileName, "ActorMap_%s", std::to_string(actorHash).c_str());
+	sprintf_s(writerFoundEvent, "WriterFoundEvent_%s", std::to_string(actorHash).c_str());
+	sprintf_s(readerFinishedEvent, "ReaderFinishedEvent_%s", std::to_string(actorHash).c_str());
+
+	m_MapFile = std::make_unique<IPCMap>(sizeof(FixedSizeQueue<ChunkInfoType, MAX_CHUNKS>), fileName);
+	m_WriterFoundEvent = std::make_unique<IPCEvent>(writerFoundEvent);
+	m_ReaderFinishedEvent = std::make_unique<IPCEvent>(readerFinishedEvent);
+	m_QueueManager = std::make_unique<IPCSafeQueueManager<ChunkInfoType, MAX_CHUNKS>>(
+		static_cast<FixedSizeQueue<ChunkInfoType, MAX_CHUNKS>*>(m_MapFile->GetHandle()), m_ActorData);
 }
 
 void Actor::Start()
 {
-	if (Role::Reader == m_eRole)
-		Read();
+	if (m_MapFile->IsCreated())
+	{
+		ReadFile();
+	}
 	else
-		Write();
-}
-
-std::string Actor::GetSrcPath() const
-{
-	return m_strSrcPath;
-}
-
-std::string Actor::GetDstPath() const
-{
-	return m_strDstPath;
-}
-
-size_t Actor::GetId() const
-{
-	return m_nId;
-}
-
-Role Actor::GetRole() const
-{
-	return m_eRole;
-}
-
-bool Actor::Finished() const
-{
-	return m_bFinished;
-}
-
-bool Actor::Ready() const
-{
-	return m_bReady;
-}
-
-void Actor::SetBufferId(size_t nBufferId)
-{
-	m_nBufferId = nBufferId;
+	{
+		WriteFile();
+	}
 }
 
-size_t Actor::GetBufferId() const
+void Actor::ReadFile()
 {
-	return m_nBufferId;
-}
+	if (WAIT_TIMEOUT == m_WriterFoundEvent->Wait(5000))
+	{
+		return;
+	}
 
-void Actor::Read()
-{
-	std::cout << "Read\n";
-	std::ifstream input(m_strSrcPath, std::ios::binary);
+	std::ifstream input(m_ActorData.SrcPath, std::ios::binary);
 
 	auto rdBuf = input.rdbuf();
 	size_t nSize = 0;
 
 	ChunkType chunk;
 
-	DWORD rt = m_condition.Wait(5000);
-
-	if (WAIT_TIMEOUT == rt)
+	while ((nSize = rdBuf->sgetn(chunk.data(), chunk.size())) && nSize)
 	{
-		m_bFinished = true;
-		std::cout << "Timed Out!\n";
-		return;
+		m_QueueManager->Push({ chunk, nSize });
 	}
 
-	while (!m_bFinished)
-	{
-		m_bFinished = !((nSize = rdBuf->sgetn(chunk.data(), chunk.size())) && nSize);
-		std::cout << "Read " << chunk.data() << std::endl;
-		m_pBuffer->Push({chunk, nSize});
-	}
-	std::cout << "Finished Read\n";
+	m_ReaderFinishedEvent->Notify();
 }
 
-void Actor::Write()
+void Actor::WriteFile()
 {
-	std::ofstream output(m_strDstPath, std::ios::binary | std::ios::trunc);
+	m_WriterFoundEvent->Notify();
 
-	size_t nReaderId = m_pBuffer->GetReaderId();
-	Actor* pReader = m_pAdministrator->GetActorPtr(nReaderId);
+	std::ofstream output(m_ActorData.DstPath, std::ios::binary | std::ios::trunc);
 
-	m_condition.NotifyOne();
-	std::cout << "Write\n";
+	bool readerFinished = false;
 
-	while (!(pReader->Finished() && m_pBuffer->Empty()))
+	while ((readerFinished = (ERROR_SUCCESS == m_ReaderFinishedEvent->Wait(0)) && !readerFinished) || !m_QueueManager->Empty())
```

Comment:
I don't see ERROR_SUCCESS in documentation https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobjectex#return-value
It's quite complicated expression which I can't check:
```
while ((readerFinished = (ERROR_SUCCESS == m_ReaderFinishedEvent->Wait(0)) && !readerFinished) || !m_QueueManager->Empty())
``` 
This is not a good codes tyle to assign a value in condition operator. Resharper highlights it as a warning. 
My version of that cycle is:
```
void Actor::WriteFile()
{
	std::cout << "Writer\n";
	IPCEvent writerFoundEvent{ MakeName(WriterFoundEventName, m_actorHash) };
	writerFoundEvent.Notify();

	IPCEvent readerFinishedEvent{ MakeName(ReaderFinishedEvent, m_actorHash) };

	std::ofstream output(m_ActorData.DstPath, std::ios::binary | std::ios::trunc);

	bool readerFinished = false;
	do
	{
		if (!m_QueueManager->Empty())
		{
			auto buffData = m_QueueManager->Pop();
			output.write(buffData.first.data(), buffData.second);
			//std::cout << "w:" << total << "\n";
		}
		const auto waitResult = readerFinishedEvent.Wait(10);
		readerFinished |= WAIT_OBJECT_0 == waitResult;
	} while (!readerFinished || !m_QueueManager->Empty());
	//writerFoundEvent.Notify();
	std::cout << "Write completed";
}
```